### PR TITLE
SNOW-983006: Copy column states, schema query and query params when copying a DataFrame

### PR DESCRIPTION
**Pull request by sfc-gh-yixie**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-983066

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

**Comment by sfc-gh-stan** [comment]
The changes LGTM. What is the context of [SNOW-983006](https://snowflakecomputing.atlassian.net/browse/SNOW-983006)?

File: src/snowflake/snowpark/dataframe.py [hunk 1]
```diff
@@ -726,7 +726,15 @@ def to_local_iterator(
         )
 
     def __copy__(self) -> "DataFrame":
-        return DataFrame(self._session, copy.copy(self._select_statement or self._plan))
+        if self._select_statement:
+            new_plan = copy.copy(self._select_statement)
+            new_plan.column_states = self._select_statement.column_states
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/dataframe.py:731]: I think these should be deep copies.

Comment (sfc-gh-stan) [reply src/snowflake/snowpark/dataframe.py:731]: We had a discussion offline and these changes ideally should be in SelectStament.__copy__ but that will be a risk change (with multiple existing references to SelectStatement.__copy__) to check in last minute. The plan is to make these changes here for now and move it to SelectStatement.__copy__ in the Feburary's release.

File: CHANGELOG.md [hunk 2]
```diff
@@ -24,6 +24,12 @@
   - Earlier timestamp columns with a timezone would be inferred as `TimestampType(TimestampTimeZone.NTZ)` and loose timezone information but will now be correctly inferred as `TimestampType(TimestampTimeZone.LTZ)` and timezone information is retained correctly.
   - Set session parameter `PYTHON_SNOWPARK_USE_LOGICAL_TYPE_FOR_CREATE_DATAFRAME` to revert back to old behavior. It is recommended that you update your code soon to align with correct behavior as the parameter will be removed in the future.
 - Fixed a bug that `DataFrame.to_pandas` gets decimal type when scale is not 0, and creates an object dtype in `pandas`. Instead, we cast the value to a float64 type.
+- Fixed a bug that aliasing a DataFrame column raises an error when the DataFame is copied from another DataFrame with an aliased column. For instance,
+  ```python
+  df = df.select(col("a").alias("b"))
```
[the remark this answers is not in the document]
Comment (sfc-gh-mabrennan) [reply CHANGELOG.md:29]: I think having the code snippet is ok.

**Review by sfc-gh-sfan — APPROVED** [review]
LGTM

**Review by sfc-gh-stan — APPROVED** [review]

**Pull request merged**